include/seqan3/container/detail/ra_iterator.hpp

```diff
@@ -0,0 +1,218 @@
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+// Authors: Marie Hoffmann <marie.hoffmann AT fu-berlin.de>
+// ============================================================================
+// Implementation of the (const) iterator given a container reference.
+// ============================================================================
+
+#pragma once
+
+#include <iomanip>
+#include <limits>
+#include <type_traits>
+
+#include <range/v3/all.hpp>
+#include <seqan3/container/concepts.hpp>
+
+namespace seqan3::detail
+{
+
+template <typename container_type, bool is_const=false>
+    requires random_access_range_concept<container_type>
+class ra_iterator
+{
+
+private:
+    // Note: an iterator can only be initialized with container (host) reference
+    container_type & host; //{*(container_type*)0};
+    // set to numeric_limits::max to indicate that iterator exceeds container size
+    typename container_type::size_type pos; //{0};
+public:
+    typedef typename container_type::size_type size_type;
+    typedef typename container_type::difference_type difference_type;
+    typedef typename container_type::value_type value_type;
+    using reference = std::conditional_t<is_const, value_type const &, value_type &>;
+    typedef typename container_type::const_reference const_reference;
+    using pointer = std::conditional_t<is_const, value_type const *, value_type *>;
+
+    // Explicitly not possible: ra_iterator() : host(*(container_type*)0) {};
```

Comment:
We need a default constructor, otherwise this iterator does not fulfil the ```iterator_concept```, which requires ```t``` to be a semi-regular type, thus it must be default constructible. 
The container must be stored as  a pointer, which is set to ```nullptr``` in the default case.